TLS for VectorPostprocessors

permcody opened the pull request:
After this is merged we can go through the existing samplers and remove temporary vectors
or other workarounds to not having TLS.

refs #7809

Ping @friedmud, @dschwen

permcody committed:
TLS for VectorPostprocessors

After this is merged we can go through the existing samplers and remove temporary vectors
or other workarounds to not having TLS.

refs #7809

permcody commented:
Exactly as they are in UserObjects and Postprocessors. We already have a `threadJoin()` routine that's being called on these objects. The content of these extra vectors is more or less unimportant. It's just there as scratch space as the VectorPostprocessor executes.

permcody commented on framework/src/vectorpostprocessors/VectorPostprocessor.C:
```diff
@@ -47,5 +48,8 @@ VectorPostprocessor::getVector(const std::string & vector_name)
 {
-  return _vpp_fe_problem->declareVectorPostprocessorVector(_vpp_name, vector_name);
+  if (_vpp_tid)
+    return _thread_local_vector;
```
Exactly as they are in UserObjects and Postprocessors. We already have a threadJoin() routine that's being called on these objects. The content of these extra vectors is more or less unimportant. It's just there as scratch space as the VectorPostprocessor executes.

friedmud replied:
There needs to be one local vector for each `vector_name`, right?

dschwen replied:
Yeah, this will not work for VPPs that declare more than one vector :-/

permcody replied:
Fixed

friedmud commented:
Also: there should be a test or two showing how to use the new functionality.

permcody commented:
Yes - we definitely do need multiple vectors. I'll also add a test object.

permcody committed:
Revert "Fix thread fail (#7801)"

This reverts commit 513c5dfae1ae1d85b1e247baf45da0566656006c.

permcody committed:
Fix sign comparison warning
Also add override keywords

permcody committed:
Rewrite SamplerBase to avoid temporaries

permcody commented:
So I decided not to add a dummy test object for this. I reverted Daniel's "fix" instead so we have a real object and a real test in the framework already with the Histogram VPP. Also, I rewrote SamplerBase to get rid of all the temporaries so we can verify that this all works with threads properly.

dschwen commented:
The revert of the fix also stripped the `override`s off

permcody commented:
Well you can't have your cake and eat it too, but apparently you "can has" all the donuts you want (and do want).

I'll add those back into the separate fixup commit that already sits on top of the revert commit and repush.

permcody commented on framework/include/vectorpostprocessors/VectorPostprocessorData.h:
```diff
@@ -64,26 +70,27 @@ class VectorPostprocessorData : public Restartable
-  std::map<std::string, std::map<std::string, VectorPostprocessorValue*> > _values_old;
+  std::set<std::string> _requested_items;
+  std::set<std::string> _supplied_items;
 };
```
Yes and no. Right now there's already dependency resolution on the objects themselves because they are UserObjects and sorted in the warehouse. The intention for this was to add in a new sanity check for the vectors themselves if you couple to vectors, there is no check on whether a vector that you couple to exists right now. I started on this but need to flesh it out quite a bit more and want to unify the way it works between VectorPostprocessors and Postprocessors. I can either continue to work on this or continue that work in another PR. I kind of want to handle it separately since this is already getting to be a substantial PR.

friedmud commented:
@permcody BTW: I'm somewhat abusing the fact that there is no sanity checking on VPP vector declarations for my own purposes.

Check out: https://github.com/friedmud/mocodile/blob/devel/src/vectorpostprocessors/RayTracingResultsVectorPostprocessor.C

I don't declare the vectors until `initialize()`.  That's because I don't actually know what they are until that point (this is outputting a `std::map<std::vector>` that is built elsewhere during execution).

Please don't undo my ability to do that 😉

permcody commented:
Ha! Well your "we allow that" may have to take a back seat to having MOOSE do something intelligent for our users. We can still support late creation of vectors with additional Actions or something. At some point though we really should be able to check the sanity of the system before beginning execution. I already got bit on this on one of my first uses of VectorPostprocessors since I renamed my vector during development and my coupled object worked just fine and drove me mad while I tried to figure out while I only had zeros in my vector (or nothing at all).

BTW - @YaqiWang abuses Postprocessors in a similar fashion so I don't have satisfactory checks like I want to.

I'm supportive of late declarations, we might just need to design that in to the system so that we can still check them.

friedmud commented:
Sounds good.  I was just throwing the idea out there.  Since there aren't
that many users of VPP's yet (but it is growing) we need to make sure to
cover our use-cases as development continues...

On Tue, Oct 11, 2016 at 2:41 PM Cody Permann <notifications@github.com>
wrote:

> Ha! Well your "we allow that" may have to take a back seat to having MOOSE
> do something intelligent for our users. We can still support late creation
> of vectors with additional Actions or something. At some point though we
> really should be able to check the sanity of the system before beginning
> execution. I already got bit on this on one of my first uses of
> VectorPostprocessors since I renamed my vector during development and my
> coupled object worked just fine and drove me mad while I tried to figure
> out while I only had zeros in my vector (or nothing at all).
>
> BTW - @YaqiWang <https://github.com/YaqiWang> abuses Postprocessors in a
> similar fashion so I don't have satisfactory checks like I want to.
>
> I'm supportive of late declarations, we might just need to design that in
> to the system so that we can still check them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7819#issuecomment-253006729>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMTd2X5zhClphcRb5lNG-LQCdIJSjks5qy9hngaJpZM4KPSQ1>
> .
>

permcody committed:
Minor refactor of PointerSamplerBase

permcody committed:
Fix broken test detected by assert

permcody committed:
Update SphericalAverage VPP to avoid temporaries

permcody commented on framework/include/vectorpostprocessors/SphericalAverage.h:
```diff
@@ -64,11 +64,8 @@ class SphericalAverage : public ElementVectorPostprocessor
-
   /// sample count per bin
-  VectorPostprocessorValue _count_tmp;
+  std::vector<unsigned int> _counts;
```
Note to reviewer: I think we should avoid creating VectoPostprocessor instances in our objects. We may eventually want to make the constructor private like we have for Material and Variable containers to help prevent users from missing the reference in normal declarations. This particular vector is not a declared vector so I switched it to C++ vector.

permcody commented:
@friedmud - care to review this?

friedmud commented:
Looking

friedmud commented on framework/src/vectorpostprocessors/PointSamplerBase.C:
```diff
@@ -100,7 +108,7 @@ void
 {
   // Save off for speed
-  unsigned int pid = processor_id();
+  const auto pid = processor_id();
```
Much better... we gotta get these `unsigned int`s out of the code... and, yes, I realize that a lot of them are my fault...

friedmud commented on framework/src/vectorpostprocessors/PointSamplerBase.C:
```diff
@@ -111,27 +119,27 @@ PointSamplerBase::finalize()
   {
-    // Only do this check on the proc zero because it's the same on every processor
-    // _found_points should contain all 1's at this point (ie every point was found by a proc)
+     // Only do this check on the proc zero because it's the same on every processor
```
Why the extra spaces here?

friedmud commented on framework/src/vectorpostprocessors/SamplerBase.C:
```diff
@@ -44,104 +44,106 @@ void
-  for (unsigned int i=0; i<variable_names.size(); i++)
-    _values[i] = &_vpp->declareVector(variable_names[i]);
+  for (const auto & variable_name : variable_names)
+    _values.push_back(&_vpp->declareVector(variable_name));
```
`emplace_back`?  I know it doesn't matter much here... but I've started using it for essentially everything...

friedmud commented on framework/src/vectorpostprocessors/SamplerBase.C:
```diff
@@ -44,104 +44,106 @@ void
-  _x_tmp.push_back(p(0));
-  _y_tmp.push_back(p(1));
-  _z_tmp.push_back(p(2));
+  _x.push_back(p(0));
```
This is better

friedmud commented on framework/src/vectorpostprocessors/SamplerBase.C:
```diff
@@ -44,104 +44,106 @@ void
-    for (unsigned int j=0; j<sorted_indices.size(); j++)
-      (*_values[i])[j] = _values_tmp[i][sorted_indices[j]];
+    // Sort by the vector chosen by the user
+    if (i == _sort_by)
```
Why not move this outside the `for` loop?

permcody replied:
wow - missed that. This will allow me to clean up the code quite a bit more (I see further collapsing cascades after this.

friedmud commented on framework/src/vectorpostprocessors/SamplerBase.C:
```diff
@@ -44,104 +44,106 @@ void
+  /**
+   * We have several vectors that all need to be processed in the same way.
+   * Rather than enumerate them all, let's just create a vector a pointers
+   * and work on them that way.
```
Damn - I got straight up _schooled_ here.

Excellent piece of coding!

This is almost a master class in how all of MOOSE should look...

andrsd replied:
A typo? 
> vector __of__ pointers

permcody replied:
Fixed

friedmud commented on framework/src/vectorpostprocessors/SamplerBase.C:
```diff
@@ -44,104 +44,106 @@ void
+#ifndef NDEBUG
+  for (const auto vec_ptr : vec_ptrs)
+    if (vec_ptr->size() != vector_length)
+      mooseError("Vector length mismatch");
```
Yeah - it would nice to be able to actually handle this... like only sort the values that exist or something.  But it's an edge case that can't happen with any of our existing objects that use this interface right now.

permcody replied:
That'll take a different data structure because we don't have an association of which items go with which ids if the lengths are different. That's a problem for a different day.

friedmud commented on framework/src/vectorpostprocessors/SamplerBase.C:
```diff
@@ -44,104 +44,106 @@ void
-    _z[i] = _z_tmp[index];
-    _id[i] = _id_tmp[index];
+    // Swap vector storage with sorted vector
+    vec_ptrs[i]->swap(tmp_vector);
```
Very nice.

friedmud commented on framework/src/vectorpostprocessors/VectorPostprocessorData.C:
```diff
@@ -29,38 +29,57 @@ VectorPostprocessorData::hasVectorPostprocessor(const std::string & name)
+VectorPostprocessorData::getVectorPostprocessorHelper(const VectorPostprocessorName & vpp_name, const std::string & vector_name, bool get_current)
+{
+  // Intentional use of RHS brackets on a std::map to do a multilevel retrieve or insert
+  auto & vec_struct = _values[vpp_name][vector_name];
```
You do it here and it's ok ;-)

permcody replied:
Doing a double layer insert is really, really ugly with `lower_bound`. It's actually more than 2*3. I tried it and thought it was atrocious.

friedmud commented on framework/src/vectorpostprocessors/VectorPostprocessorData.C:
```diff
@@ -29,38 +29,57 @@ VectorPostprocessorData::hasVectorPostprocessor(const std::string & name)
-
-  if (pp_val == NULL)
-    pp_val = &declareRestartableDataWithObjectName<VectorPostprocessorValue>(vpp_name + "_" + vector_name, "values_old");
+  _requested_items.emplace(vpp_name + "::" + vector_name);
```
This means that you will create a dependency on this VPP... even though it's only a dependency on the _old_ value of this VPP.

That can lead to erroneous cyclic dependencies (like the kind we've had to clear up elsewhere).

I think that for now it's ok to just _not_ note the dependency on the old value.  It should always be available...

permcody replied:
Nice catch, I'll fix this

permcody replied:
oh wait... This _might_ not be bad depending on how it's used. We actually need three different array, or at least one additional flag: these two for creating the sanity checking (making sure that supplied and requested are on parity), but we also need a flag or something to indicate whether or not a dependency should be created for a given pair. Since this implementation isn't finished, I'll leave it for now.

friedmud commented on framework/src/vectorpostprocessors/VectorPostprocessorData.C:
```diff
@@ -29,38 +29,57 @@ VectorPostprocessorData::hasVectorPostprocessor(const std::string & name)
+  if (!vec_struct.current)
+  {
+    mooseAssert(!vec_struct.old, "Uninitialized pointers in VectorPostprocessor Data");
+    vec_struct.current = &declareRestartableDataWithObjectName<VectorPostprocessorValue>(vpp_name + "_" + vector_name, "values");
```
Yuck - does this have to be this way for "lazy" binding?  It seems like the restartable data should be declared when the vector is declared.

(Note: I realize that the old code was doing the same)

permcody replied:
I'm sure we can do better. This `...Data` class is isolated from users so I'm not worried about fixing this right now. These things should be on the radar later for refactoring opportunities.

friedmud reviewed (changes requested):
I'm marking this as `Request changes`... but really it's ok the way it is. 

I made a few suggestions... but it's mergeable as is.

Overall this is a really great leap forward in code quality in this area.  Thanks for taking the time to do it really well!

permcody commented on framework/src/outputs/TableOutput.C:
```diff
@@ -85,18 +85,22 @@ TableOutput::outputVectorPostprocessors()
+    const auto & vectors = _problem_ptr->getVectorPostprocessorVectors(vpp_name);
 
-    FormattedTable & table = _vector_postprocessor_tables[vpp_name];
+    auto table_it = _vector_postprocessor_tables.lower_bound(vpp_name);
```
That's a fair criticism. I'm very, very tainted on RHS use of std::map::operator[]. I would say that I track down a bug about every 4-6 weeks that's related to careless use of that operator. I'd almost rather eradicate it's use anywhere in MOOSE but the syntax is very convenient which makes that hard to do. 

This particular case is actually one of the few cases where RHS brackets are OK because we want to always create tables so I'm torn. In general, this is the syntax I'd like to move to because for the normal access case, you either add a straight up error when the item isn't found or often just an assertion when you are "sure" the item is there (yeah right...)

Here's another crazy idea... We use maps a lot in MOOSE. Maybe it's time to think about replacing our use of maps with a better container. We could start with just aliasing our container to std::map but quickly move to hiding implementation details such as the code you see there behind familiar operators like brackets. I've been using this `lower_bound` code lately because it's the best way to check for an item and insert if if necessary because you can use `hint`. Using `find` doesn't work because you always end up with a iterator to the end of the map meaning you have to do two tree searches to handle the insert case. Is it premature optimization? Maybe, but maybe not. I'm starting to put these in as I fix accidental insertion.
</rant>

permcody commented on framework/src/outputs/TableOutput.C:
```diff
@@ -85,18 +85,22 @@ TableOutput::outputVectorPostprocessors()
 
     for (const auto & vec_it : vectors)
     {
-      VectorPostprocessorValue vector = *(vec_it.second);
```
Yep

permcody committed:
Address comments

permcody commented:
Comments have been addressed.

dschwen commented on framework/src/outputs/TableOutput.C:
```diff
@@ -85,18 +85,22 @@ TableOutput::outputVectorPostprocessors()
+      const auto & vector = *vec_it.second.current;
 
-      for (unsigned int i=0; i<vector.size(); i++)
+      for (auto i = beginIndex(vector); i < vector.size(); ++i)
```
I used it in 513c5dfa after you made a comment to that effect :-D. It might just take a bit to catch on. I suggest we start converting loops (if we decide to stick with it). Seeing it in action in existing code may improve adoption.

YaqiWang replied:
This looks to me should be a method in vector, vector.begin_index(). This beginIndex looks ugly although correct.

permcody replied:
Yeah, sadly there is no method for creating an enumeration. There are other options and they are all uglier.

```auto i = decltype(foo)(0)```
or if you don't have a const type
```decltype(foo) i = 0```

permcody commented:
This PR should be ready to merge, any takers?

permcody commented:
@andrsd - Would you mind reviewing this PR and merging if appropriate? It's been sitting here for awhile.

friedmud commented:
@permcody This PR breaks my application...

Like I said: I need to be able to declare vectors after the solve.  I thought this wasn't going to mess with that?

Reverting this merge commit fixes my problem...

How do you want to proceed?  This is required for my application and my immediate deliverables.  Can you see an easy path forward?

permcody commented:
I guess we need to get your App fixed and on our our testing list! I'm sure
we can accomodate the design of this PR with your needs. Are you just
getting an error because something isn't already declared?

On Fri, Nov 11, 2016 at 3:44 PM Derek Gaston <notifications@github.com>
wrote:

> @permcody <https://github.com/permcody> This PR breaks my application...
>
> Like I said: I need to be able to declare vectors after the solve. I
> thought this wasn't going to mess with that?
>
> Reverting this merge commit fixes my problem...
>
> How do you want to proceed? This is required for my application and my
> immediate deliverables. Can you see an easy path forward?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7819#issuecomment-260073019>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIEb2EspJ0VD8oQchO5Pi3PokJg-Sks5q9O_agaJpZM4KPSQ1>
> .
>

friedmud commented:
I get a segfault in opt.  In dbg I get an error about a key missing in
VectorPostprocessorData.

On Fri, Nov 11, 2016 at 7:41 PM Cody Permann <notifications@github.com>
wrote:

> I guess we need to get your App fixed and on our our testing list! I'm sure
> we can accomodate the design of this PR with your needs. Are you just
> getting an error because something isn't already declared?
>
> On Fri, Nov 11, 2016 at 3:44 PM Derek Gaston <notifications@github.com>
> wrote:
>
> > @permcody <https://github.com/permcody> This PR breaks my application...
> >
> > Like I said: I need to be able to declare vectors after the solve. I
> > thought this wasn't going to mess with that?
> >
> > Reverting this merge commit fixes my problem...
> >
> > How do you want to proceed? This is required for my application and my
> > immediate deliverables. Can you see an easy path forward?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/7819#issuecomment-260073019>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AC5XIEb2EspJ0VD8oQchO5Pi3PokJg-Sks5q9O_agaJpZM4KPSQ1
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7819#issuecomment-260088125>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMc1PN5WFW2XHwlkQZeEqRK2z0_zBks5q9QtCgaJpZM4KPSQ1>
> .
>

permcody commented:
Is that without threads? I'm surprised that anything changed for the
non-threaded case.

On Fri, Nov 11, 2016 at 5:46 PM Derek Gaston <notifications@github.com>
wrote:

> I get a segfault in opt. In dbg I get an error about a key missing in
> VectorPostprocessorData.
>
> On Fri, Nov 11, 2016 at 7:41 PM Cody Permann <notifications@github.com>
> wrote:
>
> > I guess we need to get your App fixed and on our our testing list! I'm
> sure
> > we can accomodate the design of this PR with your needs. Are you just
> > getting an error because something isn't already declared?
> >
> > On Fri, Nov 11, 2016 at 3:44 PM Derek Gaston <notifications@github.com>
> > wrote:
> >
> > > @permcody <https://github.com/permcody> This PR breaks my
> application...
> > >
> > > Like I said: I need to be able to declare vectors after the solve. I
> > > thought this wasn't going to mess with that?
> > >
> > > Reverting this merge commit fixes my problem...
> > >
> > > How do you want to proceed? This is required for my application and my
> > > immediate deliverables. Can you see an easy path forward?
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/idaholab/moose/pull/7819#issuecomment-260073019>,
> > or mute
> > > the thread
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AC5XIEb2EspJ0VD8oQchO5Pi3PokJg-Sks5q9O_agaJpZM4KPSQ1
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/7819#issuecomment-260088125>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AA1JMc1PN5WFW2XHwlkQZeEqRK2z0_zBks5q9QtCgaJpZM4KPSQ1
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7819#issuecomment-260088567>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIKyjU0t1lQgmUtBTxfOvW0b5xBr9ks5q9QxUgaJpZM4KPSQ1>
> .
>

friedmud commented:
Yep - without threads.

On Fri, Nov 11, 2016 at 8:13 PM Cody Permann <notifications@github.com>
wrote:

> Is that without threads? I'm surprised that anything changed for the
> non-threaded case.
>
> On Fri, Nov 11, 2016 at 5:46 PM Derek Gaston <notifications@github.com>
> wrote:
>
> > I get a segfault in opt. In dbg I get an error about a key missing in
> > VectorPostprocessorData.
> >
> > On Fri, Nov 11, 2016 at 7:41 PM Cody Permann <notifications@github.com>
> > wrote:
> >
> > > I guess we need to get your App fixed and on our our testing list! I'm
> > sure
> > > we can accomodate the design of this PR with your needs. Are you just
> > > getting an error because something isn't already declared?
> > >
> > > On Fri, Nov 11, 2016 at 3:44 PM Derek Gaston <notifications@github.com
> >
> > > wrote:
> > >
> > > > @permcody <https://github.com/permcody> This PR breaks my
> > application...
> > > >
> > > > Like I said: I need to be able to declare vectors after the solve. I
> > > > thought this wasn't going to mess with that?
> > > >
> > > > Reverting this merge commit fixes my problem...
> > > >
> > > > How do you want to proceed? This is required for my application and
> my
> > > > immediate deliverables. Can you see an easy path forward?
> > > >
> > > > —
> > > > You are receiving this because you were mentioned.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/idaholab/moose/pull/7819#issuecomment-260073019
> >,
> > > or mute
> > > > the thread
> > > > <
> > >
> >
> https://github.com/notifications/unsubscribe-auth/AC5XIEb2EspJ0VD8oQchO5Pi3PokJg-Sks5q9O_agaJpZM4KPSQ1
> > > >
> > > > .
> > > >
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/idaholab/moose/pull/7819#issuecomment-260088125>,
> > or mute
> > > the thread
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AA1JMc1PN5WFW2XHwlkQZeEqRK2z0_zBks5q9QtCgaJpZM4KPSQ1
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/7819#issuecomment-260088567>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AC5XIKyjU0t1lQgmUtBTxfOvW0b5xBr9ks5q9QxUgaJpZM4KPSQ1
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7819#issuecomment-260090900>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMSn454yAd-wCbYJ22XWHQN7Oei0gks5q9RKjgaJpZM4KPSQ1>
> .
>

friedmud commented:
If you want to run it:

https://github.com/friedmud/mocodile

The test is:

deterministic/parameter_study

friedmud commented:
Here's what I get in `opt`... I get a segfault here:

```
* thread #1: tid = 0x30abfa, 0x0000000100672a00 libmoose-opt.0.dylib`TableOutput::outputVectorPostprocessors() + 592, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x38)
  * frame #0: 0x0000000100672a00 libmoose-opt.0.dylib`TableOutput::outputVectorPostprocessors() + 592
    frame #1: 0x0000000100645d5e libmoose-opt.0.dylib`CSV::outputVectorPostprocessors() + 14
    frame #2: 0x000000010062eee9 libmoose-opt.0.dylib`AdvancedOutput<FileOutput>::output(ExecFlagType const&) + 1257
    frame #3: 0x0000000100645daf libmoose-opt.0.dylib`CSV::output(ExecFlagType const&) + 63
    frame #4: 0x000000010066b12c libmoose-opt.0.dylib`OutputWarehouse::outputStep(ExecFlagType) + 76
    frame #5: 0x000000010045bceb libmoose-opt.0.dylib`MooseApp::executeExecutioner() + 75
    frame #6: 0x000000010045cab6 libmoose-opt.0.dylib`MooseApp::run() + 118
    frame #7: 0x0000000100001522 mocodile-opt`main + 114
    frame #8: 0x00007fff970dc5ad libdyld.dylib`start + 1
    frame #9: 0x00007fff970dc5ad libdyld.dylib`start + 1
```

friedmud commented:
In dbg I get this:

```
Assertion `vec_pair != _values.end()' failed
No vectors found for vpp_name: ray_tracing
at /Users/gastdr/projects/moose/framework/src/vectorpostprocessors/VectorPostprocessorData.C, line 74
Stack frames: 15
0: 0   libmesh_dbg.0.dylib                 0x000000010454a08f libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 2287
1: 1   libmoose-dbg.0.dylib                0x00000001035fcab9 VectorPostprocessorData::vectors(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const + 2457
2: 2   libmoose-dbg.0.dylib                0x00000001026653df FEProblem::getVectorPostprocessorVectors(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 47
3: 3   libmoose-dbg.0.dylib                0x0000000103126cc5 TableOutput::outputVectorPostprocessors() + 581
4: 4   libmoose-dbg.0.dylib                0x00000001030a466f CSV::outputVectorPostprocessors() + 31
5: 5   libmoose-dbg.0.dylib                0x0000000103068204 AdvancedOutput<FileOutput>::output(ExecFlagType const&) + 2052
6: 6   libmoose-dbg.0.dylib                0x00000001030a46dd CSV::output(ExecFlagType const&) + 93
7: 7   libmoose-dbg.0.dylib                0x0000000103017cc6 AdvancedOutput<FileOutput>::outputStep(ExecFlagType const&) + 182
8: 8   libmoose-dbg.0.dylib                0x000000010310a4d5 OutputWarehouse::outputStep(ExecFlagType) + 453
9: 9   libmoose-dbg.0.dylib                0x0000000102670075 FEProblem::outputStep(ExecFlagType) + 101
10: 10  libmoose-dbg.0.dylib                0x0000000102d110dd Steady::init() + 157
11: 11  libmoose-dbg.0.dylib                0x0000000102a94bd4 MooseApp::executeExecutioner() + 180
12: 12  libmoose-dbg.0.dylib                0x0000000102a9733c MooseApp::run() + 140
13: 13  mocodile-dbg                        0x0000000101c8f3df main + 303
14: 14  libdyld.dylib                       0x00007fff970dc5ad start + 1
```

friedmud commented:
That last one is spot on... there _are_ no vectors for that VPP... because it hasn't declared any yet (this is `initial` output)

friedmud commented:
The basic issue is that `VectorPostprocessor::vectors()` and `FEProbem::getVectorPostprocessorVectors()` both return _references_... but if there are no vectors there is nothing to return.

friedmud commented:
Trying to fix it - stay tuned
